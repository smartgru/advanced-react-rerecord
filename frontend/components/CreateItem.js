import React, { Component } from 'react';
import { Mutation, useMutation } from '@apollo/client';

import gql from 'graphql-tag';
import Router from 'next/router';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createItem(
      data: {
        name: $name
        description: $description
        price: $price
        image: $image
      }
    ) {
      id
    }
  }
`;

function CreateItem() {
  const { inputs, handleChange } = useForm({
    name: 'Nice Shoes',
    description: 'soo nice',
    image: '',
    price: 500,
  });

  const [createItem, { loading, error }] = useMutation(CREATE_ITEM_MUTATION, {
    variables: inputs,
  });

  return (
    <Form
      data-testid="form"
      onSubmit={async e => {
        // Stop the form from submitting
        e.preventDefault();
        // call the mutation
        const res = await createItem();
        // change them to the single item page
        Router.push({
          pathname: '/item',
          query: { id: res.data.createItem.id },
        });
      }}
    >
      <Error error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="file">
          Image
          <input
            type="file"
            id="file"
            name="image"
            placeholder="Upload an image"
            required
            onChange={handleChange}
          />
        </label>

        <label htmlFor="name">
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Name"
            required
            value={inputs.name}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            required
            value={inputs.price}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Description"
            required
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Submit</button>
      </fieldset>
    </Form>
  );
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
