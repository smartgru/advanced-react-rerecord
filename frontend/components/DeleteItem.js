import React, { Component } from 'react';
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

// TODO: Update the cache in Apollo 3
const update = () => {
  console.log('You didnt write this code yet');
};
// update = (cache, payload) => {
//   // manually update the cache on the client, so it matches the server
//   // 1. Read the cache for the items we want
//   const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
//   console.log(data, payload);
//   // 2. Filter the deleted itemout of the page
//   data.items = data.items.filter(
//     item => item.id !== payload.data.deleteItem.id
//   );
//   // 3. Put the items back!
//   cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
// };

function DeleteItem({ id, children }) {
  const [deleteItem, { error }] = useMutation(DELETE_ITEM_MUTATION, {
    variables: { id },
    update,
  });
  return (
    <button
      type="button"
      onClick={() => {
        if (confirm('Are you sure you want to delete this item?')) {
          deleteItem().catch(err => {
            alert(err.message);
          });
        }
      }}
    >
      {children}
    </button>
  );
}

export default DeleteItem;
