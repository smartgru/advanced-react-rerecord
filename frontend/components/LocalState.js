import { createContext, useState, useContext } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  return (
    <LocalStateProvider value={{ cartOpen, toggleCart }}>
      {children}
    </LocalStateProvider>
  );
}

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, LocalStateContext, useCart };
