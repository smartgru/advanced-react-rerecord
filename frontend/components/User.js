import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { useQuery } from '@apollo/react-hooks';
import { useState, useEffect } from 'react';

const CURRENT_USER_QUERY = gql`
  query {
    authenticatedUser {
      id
      email
      name
      permissions
      # orders {
      #   id
      # }
      cart {
        id
        quantity
        item {
          id
          price
          image {
            publicUrlTransformed
          }
          name
          description
        }
      }
    }
  }
`;

function useUser() {
  const { data } = useQuery(CURRENT_USER_QUERY);
  if (data) {
    return data.authenticatedUser;
  }
}

export default <p>Hey</p>;
export { CURRENT_USER_QUERY, useUser };
