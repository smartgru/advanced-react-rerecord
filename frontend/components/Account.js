import PropTypes from 'prop-types';
import React from 'react';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useUser } from './User';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import PleaseSignIn from './PleaseSignIn';
import FormItem from './FormItem';

const UPDATE_USER_MUTATION = gql`
  mutation UPDATE_USER_MUTATION($name: String!, $id: ID!) {
    updateUser(id: $id, data: { name: $name }) {
      id
      name
    }
  }
`;

function Account() {
  const me = useUser();
  const { inputs, handleChange } = useForm({
    name: me.name,
  });
  console.log(me);
  const [updateUser, { data, error, loading }] = useMutation(
    UPDATE_USER_MUTATION,
    {
      variables: {
        id: me.id,
        name: inputs.name,
      },
    }
  );
  return (
    <Form
      onSubmit={async e => {
        e.preventDefault();
        await updateUser();
        console.log(data);
      }}
    >
      <fieldset disabled={loading}>
        <FormItem inputs={inputs} onChange={handleChange} name="name" />
        <button type="submit">Updat{loading ? 'ing' : 'e'}</button>
      </fieldset>
    </Form>
  );
}

function Component() {
  return (
    <PleaseSignIn>
      <Account></Account>
    </PleaseSignIn>
  );
}

export default Component;
