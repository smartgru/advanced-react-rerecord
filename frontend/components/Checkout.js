import React, { useState, useEffect, useRef } from 'react';
import NProgress from 'nprogress';
import {
  CardElement,
  CardCvcElement,
  CardNumberElement,
  CardExpiryElement,
  Elements,
  useStripe,
  useElements,
} from '@stripe/react-stripe-js';
import { loadStripe } from '@stripe/stripe-js';
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import Head from 'next/head';
import Router from 'next/router';
import { CURRENT_USER_QUERY, useUser } from './User';

const CREATE_ORDER_MUTATION = gql`
  mutation checkout($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const style = {
  base: {
    color: '#32325d',
    border: '5px solid red',
    fontFamily: '"Helvetica Neue", Helvetica, sans-serif',
    fontSmoothing: 'antialiased',
    fontSize: '16px',
    '::placeholder': {
      color: '#aab7c4',
    },
  },
  invalid: {
    color: '#fa755a',
    iconColor: '#ffc600',
  },
};

function Checkout() {
  // We use loadStripe because is load in their lib async
  const stripe = loadStripe('pk_test_Vtknn6vSdcZWSG2JWvEiWSqC');
  return (
    <>
      <Elements stripe={stripe}>
        <CheckoutForm />
      </Elements>
    </>
  );
}

function useCheckout() {
  const stripe = useStripe();
  const [error, setError] = useState();
  const elements = useElements();

  const [checkout] = useMutation(CREATE_ORDER_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  // manually call the mutation once we have the stripe token

  const handleSubmit = async event => {
    // 1. Stop the form from submitting
    event.preventDefault();
    // 2. Start the page transition so show the user something is happening
    NProgress.start();
    // 3. Create the payment method
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });

    console.log({ error, paymentMethod });

    if (error) {
      return setError(error);
    }

    // 4. Send it to the server and charge it
    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    }).catch(err => {
      alert(err.message);
    });

    Router.push({
      pathname: '/order',
      query: { id: order.data.checkout.id },
    });
  };

  return { error, handleSubmit };
}

function CheckoutForm() {
  const { handleSubmit, error } = useCheckout();
  return (
    <form onSubmit={handleSubmit}>
      {error && <p>{error.message}</p>}
      <label htmlFor="cardNumber">
        Credit Card Number
        {/* <CardNumberElement /> */}
      </label>
      <CardElement />
      {/* <CardCvcElement /> */}
      <button type="submit">Pay</button>
    </form>
  );
}

export default Checkout;
