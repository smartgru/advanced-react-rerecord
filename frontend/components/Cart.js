import { useCart } from './LocalState';
import { useUser } from './User';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import CartItem from './CartItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import TakeMyMoney from './TakeMyMoney';
import Checkout from './Checkout';

function Cart() {
  const me = useUser();
  const { cartOpen, toggleCart } = useCart();
  if (!me) return null;
  return (
    <CartStyles open={cartOpen}>
      <header>
        <CloseButton onClick={toggleCart} title="close">
          &times;
        </CloseButton>
        <Supreme>{me.name}'s Cart</Supreme>
        <p>
          You Have {me.cart.length} Item{me.cart.length === 1 ? '' : 's'} in
          your cart.
        </p>
      </header>
      <ul>
        {me.cart.map(cartItem => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
        {/* <Checkout /> */}
        {me.cart.length && (
          <>
            <TakeMyMoney>
              <SickButton>Checkout</SickButton>
            </TakeMyMoney>
          </>
        )}
      </footer>
    </CartStyles>
  );
}

export default Cart;
