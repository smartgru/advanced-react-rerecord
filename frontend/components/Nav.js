import Link from 'next/link';
import { Mutation } from 'react-apollo';
import { useQuery } from '@apollo/react-hooks';
import { TOGGLE_CART_MUTATION } from './Cart';
import { CURRENT_USER_QUERY } from './User';
import NavStyles from './styles/NavStyles';

import CartCount from './CartCount';
import Signout from './Signout';

function Nav() {
  const { data: { me } = {} } = useQuery(CURRENT_USER_QUERY);
  return (
    <NavStyles data-test="nav">
      <Link href="/items">
        <a>Shop</a>
      </Link>
      {me && (
        <>
          <Link href="/sell">
            <a>Sell</a>
          </Link>
          <Link href="/orders">
            <a>Orders</a>
          </Link>
          <Link href="/me">
            <a>Account</a>
          </Link>
          <Signout />
          <Mutation mutation={TOGGLE_CART_MUTATION}>
            {toggleCart => (
              <button onClick={toggleCart}>
                My Cart
                <CartCount
                  count={me.cart.reduce(
                    (tally, cartItem) => tally + cartItem.quantity,
                    0
                  )}
                ></CartCount>
              </button>
            )}
          </Mutation>
        </>
      )}
      {!me && (
        <Link href="/signup">
          <a>Sign In</a>
        </Link>
      )}
    </NavStyles>
  );
}

export default Nav;
