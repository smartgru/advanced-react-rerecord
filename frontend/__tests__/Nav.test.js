import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { CartStateProvider } from '../components/LocalState';

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedUser: fakeUser() } },
  },
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedUser: {
          ...fakeUser(),
          cart: [fakeCartItem()],
        },
      },
    },
  },
];

describe('<Nav/>', () => {
  it('renders a minimal nav when signed out', async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={notSignedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders full nav when signed in', async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('Sign Out');
    expect(container).toHaveTextContent('My Cart');
  });

  it('renders the amount of items in the cart', async () => {
    render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocksWithCartItems}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );
    await screen.findByText('Account');
    expect(screen.getByText('3')).toBeInTheDocument();
  });
});
