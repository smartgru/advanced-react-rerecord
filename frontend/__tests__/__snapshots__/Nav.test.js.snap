// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 eEiNI"
    data-testid="nav"
  >
    <a
      href="/items"
    >
      Shop
    </a>
    <a
      href="/signup"
    >
      Sign In
    </a>
  </ul>
</div>
`;

exports[`<Nav/> renders full nav when signed in 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 eEiNI"
    data-testid="nav"
  >
    <a
      href="/items"
    >
      Shop
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/me"
    >
      Account
    </a>
    <button
      type="button"
    >
      Sign Out
    </button>
    <button
      type="button"
    >
      My Cart
      <span
        class="CartCount__AnimationStyles-xxvp4g-0 hGREFj"
      >
        <div>
          <div
            class="CartCount__Dot-xxvp4g-1 ffCQwv count"
          >
            0
          </div>
        </div>
      </span>
    </button>
  </ul>
</div>
`;
